lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -0,0 +1,611 @@
+        emitCodeForSymbolicValue(propertyVal, propertyType, builder, loc);
+    auto *newStructInst = builder.createStruct(
+        loc, expectedType, ArrayRef<SILValue>(newPropertySIL));
+    return newStructInst;
+  }
+  default: {
+    assert(false && "Symbolic value kind is not supported");
+  }
+  }
+}
+
+bool OSLogOptimization::detectAndDiagnoseErrors(
+    Optional<SymbolicValue> errorInfo, SingleValueInstruction *osLogMessageAddr,
+    ConstExprStepEvaluator &constantEvaluator) {
+
+  auto loc = osLogMessageAddr->getLoc();
+  auto sourceLoc = loc.getSourceLoc();
+  auto *fn = osLogMessageAddr->getFunction();
+  auto &module = fn->getModule();
+  auto &astContext = fn->getASTContext();
+  bool errorDetected = false;
+
+  // If we have errorInfo that indicates a fail-stop error, diagnose it.
+  if (errorInfo && constantEvaluator.isFailStopError(*errorInfo)) {
+    assert(errorInfo->getKind() == SymbolicValue::Unknown);
+    diagnose(astContext, sourceLoc, diag::oslog_const_evaluation_error);
+    errorInfo->emitUnknownDiagnosticNotes(loc);
+    errorDetected = true;
+  }
+
+  // Check if the OSLogMessage and OSLogInterpolation instances are correctly
+  // inferred as constants. If not, it implies incorrect implementation
+  // of the os log API in the overlay. Diagnostics here are for os log
+  // library authors.
+  auto osLogMessageAddrValueOpt =
+      constantEvaluator.lookupConstValue(osLogMessageAddr);
+  assert(osLogMessageAddrValueOpt.hasValue() &&
+         osLogMessageAddrValueOpt->getKind() == SymbolicValue::Address);
+
+  SmallVector<unsigned, 2> accessPath;
+  SymbolicValue osLogMessageValue =
+      osLogMessageAddrValueOpt->getAddressValue(accessPath)->getValue();
+  if (!osLogMessageValue.isConstant()) {
+    diagnose(astContext, sourceLoc, diag::oslog_non_constant_message);
+    return true;
+  }
+
+  auto osLogInterpolationValue =
+      osLogMessageValue.lookThroughSingleElementAggregates();
+  if (!osLogInterpolationValue.isConstant()) {
+    diagnose(astContext, sourceLoc, diag::oslog_non_constant_interpolation);
+    return true;
+  }
+
+  // Check if every proprety of the OSLogInterpolation instance that is a
+  // string or integer has a constant value. If this is violated this could
+  // be an indication of an error in the usage of the API. Diagnostics emitted
+  // here are for the users of the os log APIs.
+  SILType osLogMessageType = osLogMessageAddr->getType();
+  auto *structDecl = osLogMessageType.getStructOrBoundGenericStruct();
+  assert(structDecl);
+
+  VarDecl *interpolationPropDecl = structDecl->getStoredProperties().front();
+  SILType osLogInterpolationType =
+      osLogMessageType.getFieldType(interpolationPropDecl, module);
+  auto *interpolationStruct =
+      osLogInterpolationType.getStructOrBoundGenericStruct();
+  assert(interpolationStruct);
+
+  auto propertyDecls = interpolationStruct->getStoredProperties();
+  auto propertyValues = osLogInterpolationValue.getAggregateValue();
```

Comment:
Overuse of `auto` makes the code harder for me to follow. I only use auto in three cases:
(1) when the type name is already spelled (or mostly spelled) on the same line e.g. cast, getLoc, getFunction, getModule...
(2) when defining iterators that have long dependent type names
(3) when it is otherwise brutally obvious

Reply:
Sure, will fix this.